src/imgui_renderer.rs

```diff
@@ -0,0 +1,495 @@
+            vertex_buffers: &[wgpu::VertexBufferDescriptor {
+                stride: u64::try_from(Self::DRAW_VERT_SIZE).unwrap(),
+                step_mode: wgpu::InputStepMode::Vertex,
+                attributes: &[
+                    wgpu::VertexAttributeDescriptor {
+                        offset: 0,
+                        format: wgpu::VertexFormat::Float2,
+                        shader_location: 0,
+                    },
+                    wgpu::VertexAttributeDescriptor {
+                        offset: 8,
+                        format: wgpu::VertexFormat::Float2,
+                        shader_location: 1,
+                    },
+                    wgpu::VertexAttributeDescriptor {
+                        offset: 16,
+                        format: wgpu::VertexFormat::Uint,
+                        shader_location: 2,
+                    },
+                ],
+            }],
+            sample_count: 1,
+        });
+
+        // Create the font atlas texture
+
+        let mut fonts = imgui.fonts();
+        let font_texture = fonts.build_rgba32_texture();
+        let texture = device.create_texture(&wgpu::TextureDescriptor {
+            size: wgpu::Extent3d {
+                width: font_texture.width,
+                height: font_texture.height,
+                depth: 1,
+            },
+            array_layer_count: 1,
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: wgpu::TextureFormat::Rgba8Unorm,
+            usage: wgpu::TextureUsage::SAMPLED,
+        });
+
+        upload_texture_immediate(
+            device,
+            &texture,
+            font_texture.width,
+            font_texture.height,
+            font_texture.data,
+        );
+
+        let sampler = create_sampler(device);
+        let pair = Texture::new(device, &texture_layout, texture, sampler);
+        let mut textures = imgui::Textures::new();
+        let atlas_id = textures.insert(pair);
+        fonts.tex_id = atlas_id;
+
+        Ok(ImguiRenderer {
+            render_pipeline,
+            uniform_buffer,
+            uniform_bind_group,
+            texture_layout,
+            textures,
+            clear_color,
+        })
+    }
+
+    pub fn add_rgba32_texture(
+        &mut self,
+        device: &mut wgpu::Device,
+        width: u32,
+        height: u32,
+        data: &[u8],
+    ) -> imgui::TextureId {
+        let texture = device.create_texture(&wgpu::TextureDescriptor {
+            size: wgpu::Extent3d {
+                width,
+                height,
+                depth: 1,
+            },
+            array_layer_count: 1,
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: wgpu::TextureFormat::Rgba8Unorm,
+            usage: wgpu::TextureUsage::SAMPLED,
```

Comment:
You are missing the `TextureUsage::TRANSFER_DST` bit, VK validation layers complain:

```
[2019-08-19T14:12:27Z ERROR gfx_backend_vulkan]
VALIDATION [VUID-vkCmdCopyBufferToImage-dstImage-00177 (0)] : Invalid usage flag for VkImage 0x41[] used by vkCmdCopyBufferToImage(). In this case, VkImage should have VK_IMAGE_USAGE_TRANSFER_DST_BIT set during creation. The Vulkan spec states: dstImage must have been created with VK_IMAGE_USAGE_TRANSFER_DST_BIT usage flag (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkCmdCopyBufferToImage-dstImage-00177)
object info: (type: IMAGE, hndl: 65)

[2019-08-19T14:12:27Z ERROR gfx_backend_vulkan]
VALIDATION [VUID-VkImageMemoryBarrier-oldLayout-01213 (0)] : vkCmdPipelineBarrier(): Image barrier 0x000000832B8FC5C8 new Layout=VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL is not compatible with VkImage 0x41[] usage flags 0x4. The Vulkan spec states: If either oldLayout or newLayout is VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL then image must have been created with VK_IMAGE_USAGE_TRANSFER_DST_BIT set (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkImageMemoryBarrier-oldLayout-01213)
object info: (type: IMAGE, hndl: 65)

[2019-08-19T14:12:27Z ERROR gfx_backend_vulkan]
VALIDATION [VUID-VkImageMemoryBarrier-oldLayout-01213 (0)] : vkCmdPipelineBarrier(): Image barrier 0x000000832B8FCD48 old Layout=VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL is not compatible with VkImage 0x41[] usage flags 0x4. The Vulkan spec states: If either oldLayout or newLayout is VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL then image must have been created with VK_IMAGE_USAGE_TRANSFER_DST_BIT set (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkImageMemoryBarrier-oldLayout-01213)
object info: (type: IMAGE, hndl: 65)
```